### ISSUE TITLE
Residual object after switching between branches

### ISSUE THREAD

**Issue by theoryshaw** [brunopostle/ifc-git#11]
residual object (window) after switching between branches

https://gitlab.com/openingdesign/clothing_boutique

video: https://www.dropbox.com/s/3wpyewaau8hpf89/2023-04-07_22-22-31_Blender_blender.mp4?dl=0

**Comment by theoryshaw** [comment]
sometimes it crashes: 
https://www.dropbox.com/s/nl0gkaugpkwghgf/2023-04-07_22-38-12_Blender_blender.mp4?dl=0

**Comment by theoryshaw** [comment]
created a snapshot fork, for testing, if you need it.
https://gitlab.com/openingdesign/clothing-boutique-fork-for-testing

**Comment by brunopostle** [comment]
> residual object (window) after switching between branches
> 
> https://gitlab.com/openingdesign/clothing_boutique
> 
> video: https://www.dropbox.com/s/3wpyewaau8hpf89/2023-04-07_22-22-31_Blender_blender.mp4?dl=0

The 'switch revision' code deletes all the Ifc* collections and wipes `IfcStore` before reloading the IFC. I think the problem is that your model has objects outside the IfcProject heirarchy (ie. not in a blender collection):

![image](https://user-images.githubusercontent.com/1747315/230706713-f88347ad-9a89-418c-81e8-5ff63760b972.png)

This is a bug that can be fixed, but the immediate fix is to put your windows and doors in the right place and create a valid IFC.

**Comment by brunopostle** [comment]
I can reproduce the segfault, it *may* be related to the not-deleted door, will investigate. Here's the stacktrace for reference:

```
                Module blender with build-id 53a461c3e0c08a95ee7c2a17c33a870e17e16a63
                Stack trace of thread 1043086:
                #0  0x00007f32db7b25ee dict_traverse (libpython3.11.so.1.0 + 0x1b25ee)
                #1  0x00007f32db7b2004 deduce_unreachable (libpython3.11.so.1.0 + 0x1b2004)
                #2  0x00007f32db7b1662 gc_collect_main (libpython3.11.so.1.0 + 0x1b1662)
                #3  0x00007f32db84556b gc_collect_with_callback (libpython3.11.so.1.0 + 0x24556b)
                #4  0x00007f32db7a6cbd _PyObject_GC_New (libpython3.11.so.1.0 + 0x1a6cbd)
                #5  0x00007f32db7a76c4 PyDict_New (libpython3.11.so.1.0 + 0x1a76c4)
                #6  0x00007f32a2a0ce5e scan_once_unicode (_json.cpython-311-x86_64-linux-gnu.so + 0x4e5e)
                #7  0x00007f32a2a0da01 scan_once_unicode (_json.cpython-311-x86_64-linux-gnu.so + 0x5a01)
                #8  0x00007f32a2a0d0b0 scan_once_unicode (_json.cpython-311-x86_64-linux-gnu.so + 0x50b0)
                #9  0x00007f32a2a0d0b0 scan_once_unicode (_json.cpython-311-x86_64-linux-gnu.so + 0x50b0)
                #10 0x00007f32a2a0c822 scanner_call (_json.cpython-311-x86_64-linux-gnu.so + 0x4822)
                #11 0x00007f32db7ba593 _PyObject_MakeTpCall (libpython3.11.so.1.0 + 0x1ba593)
                #12 0x00007f32db7c31a4 _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c31a4)
                #13 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #14 0x00007f32db7e975b _PyVectorcall_Call (libpython3.11.so.1.0 + 0x1e975b)
                #15 0x00007f32db7c738b _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c738b)
                #16 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #17 0x00007f32db7bd2df _PyObject_FastCallDictTstate (libpython3.11.so.1.0 + 0x1bd2df)
                #18 0x00007f32db7e7ecc slot_tp_init (libpython3.11.so.1.0 + 0x1e7ecc)
                #19 0x00007f32db7ba78b type_call (libpython3.11.so.1.0 + 0x1ba78b)
                #20 0x00007f32db7ba593 _PyObject_MakeTpCall (libpython3.11.so.1.0 + 0x1ba593)
                #21 0x00007f32db7c31a4 _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c31a4)
                #22 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #23 0x000055fcb5c2b132 bpy_class_call (blender + 0x12d6132)
                #24 0x000055fcb5bc3be5 rna_operator_execute_cb (blender + 0x126ebe5)
                #25 0x000055fcb574b247 _ZL18wm_operator_invokeP8bContextP14wmOperatorTypePK7wmEventP10PointerRNAP10ReportListbb.part.0.lto_priv.0 (blender + 0xdf6247)
                #26 0x000055fcb5c270fc pyop_call.lto_priv.0 (blender + 0x12d20fc)
                #27 0x00007f32db7d8198 cfunction_call (libpython3.11.so.1.0 + 0x1d8198)
                #28 0x00007f32db7ba593 _PyObject_MakeTpCall (libpython3.11.so.1.0 + 0x1ba593)
                #29 0x00007f32db7c31a4 _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c31a4)
                #30 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #31 0x00007f32db7bd2df _PyObject_FastCallDictTstate (libpython3.11.so.1.0 + 0x1bd2df)
                #32 0x00007f32db7e80ac _PyObject_Call_Prepend (libpython3.11.so.1.0 + 0x1e80ac)
                #33 0x00007f32db88f672 slot_tp_call (libpython3.11.so.1.0 + 0x28f672)
                #34 0x00007f32db7ba593 _PyObject_MakeTpCall (libpython3.11.so.1.0 + 0x1ba593)
                #35 0x00007f32db7c31a4 _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c31a4)
                #36 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #37 0x000055fcb5c2b132 bpy_class_call (blender + 0x12d6132)
                #38 0x000055fcb5bc3be5 rna_operator_execute_cb (blender + 0x126ebe5)
                #39 0x000055fcb574b247 _ZL18wm_operator_invokeP8bContextP14wmOperatorTypePK7wmEventP10PointerRNAP10ReportListbb.part.0.lto_priv.0 (blender + 0xdf6247)
                #40 0x000055fcb5c270fc pyop_call.lto_priv.0 (blender + 0x12d20fc)
                #41 0x00007f32db7d8198 cfunction_call (libpython3.11.so.1.0 + 0x1d8198)
                #42 0x00007f32db7ba593 _PyObject_MakeTpCall (libpython3.11.so.1.0 + 0x1ba593)
                #43 0x00007f32db7c31a4 _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c31a4)
                #44 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #45 0x00007f32db7bd351 _PyObject_FastCallDictTstate (libpython3.11.so.1.0 + 0x1bd351)
                #46 0x00007f32db7e80ac _PyObject_Call_Prepend (libpython3.11.so.1.0 + 0x1e80ac)
                #47 0x00007f32db88f672 slot_tp_call (libpython3.11.so.1.0 + 0x28f672)
                #48 0x00007f32db7ba593 _PyObject_MakeTpCall (libpython3.11.so.1.0 + 0x1ba593)
                #49 0x00007f32db7c31a4 _PyEval_EvalFrameDefault (libpython3.11.so.1.0 + 0x1c31a4)
                #50 0x00007f32db7befca _PyEval_Vector (libpython3.11.so.1.0 + 0x1befca)
                #51 0x000055fcb5c2b132 bpy_class_call (blender + 0x12d6132)
                #52 0x000055fcb5bc3be5 rna_operator_execute_cb (blender + 0x126ebe5)
                #53 0x000055fcb574b247 _ZL18wm_operator_invokeP8bContextP14wmOperatorTypePK7wmEventP10PointerRNAP10ReportListbb.part.0.lto_priv.0 (blender + 0xdf6247)
                #54 0x000055fcb57116ce _ZL25wm_operator_call_internalP8bContextP14wmOperatorTypeP10PointerRNAP10ReportList21wmOperatorCallContextbPK7wmEvent.lto_priv.0 (blender + 0xdbc6ce)
                #55 0x000055fcb79710ee WM_operator_name_call_ptr_with_depends_on_cursor.constprop.0 (blender + 0x301c0ee)
                #56 0x000055fcb632025c ui_apply_but_funcs_after.lto_priv.0 (blender + 0x19cb25c)
                #57 0x000055fcb634d9a8 ui_handler_region_menu (blender + 0x19f89a8)
                #58 0x000055fcb574d4ae _ZL21wm_handlers_do_internP8bContextP8wmWindowP7wmEventP8ListBase (blender + 0xdf84ae)
                #59 0x000055fcb57509dd _ZL14wm_handlers_doP8bContextP7wmEventP8ListBase (blender + 0xdfb9dd)
                #60 0x000055fcb575614b wm_event_do_handlers (blender + 0xe0114b)
                #61 0x000055fcb571e3f2 WM_main (blender + 0xdc93f2)
                #62 0x000055fcb5190dff main (blender + 0x83bdff)
                #63 0x00007f32db047510 __libc_start_call_main (libc.so.6 + 0x27510)

**Comment by theoryshaw** [comment]
> but the immediate fix is to put your windows and doors in the right place and create a valid IFC.

That was it.  Silly me, for missing that. :)

**Comment by theoryshaw** [comment]
So i corrected the spatial tree problem. 

...

In the 'testy' branch i moved the window. 
In the 'master' branch @brunoperdigao deleted this same window.

I tried to merge the testy branch into the master branch, but the `IFC Merge failed`. 

Is there currently any conflict resolution yet?

I updated the testing fork
https://gitlab.com/openingdesign/clothing-boutique-fork-for-testing

**Comment by brunopostle** [comment]
@theoryshaw if you both modify/delete the same element in different ways then you are going to get a conflict that can't be resolved automatically (it should be ok if you both delete it or modify it in the same way).

So you can try also deleting the window in your branch, if it is deleted in both branches then it *ought* to merge.

ifcmerge outputs a list showing exactly what couldn't be merged, but this doesn't display in the IFC Git UI yet, at the moment you have to guess which elements are offending (bug #6).

**Comment by theoryshaw** [comment]
Sounds good.  I figured, just wanted to test, to make sure.

**Comment by brunopostle** [comment]
Closing this issue as it is fixed in BlenderBIM.

The crashing also appears to be fixed in BlenderBIM since [75a37cf](https://github.com/IfcOpenShell/IfcOpenShell/commit/75a37cf737a50c596986896cf465a1e0e71570d5) (I switched revision twenty or so times before it crashed, which seems pretty good considering)
